### ISSUE TITLE
Azure: Updating `loadBalancerSourceRanges` does not update the Azure NSG

### ISSUE THREAD

**Issue by rocketraman** [kubernetes/kubernetes#43608]
What keywords did you search in Kubernetes issues before filing this one?

loadBalancerSourceRanges

---

**Is this a BUG REPORT or FEATURE REQUEST?** Bug

**Kubernetes version** (use `kubectl version`): 

```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.4", GitCommit:"7243c69eb523aa4377bce883e7c0dd76b84709a1", GitTreeState:"clean", BuildDate:"2017-03-07T23:53:09Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3", GitCommit:"029c3a408176b55c30846f0faedf56aae5992e9b", GitTreeState:"clean", BuildDate:"2017-02-15T06:34:56Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: Azure
- **OS** (e.g. from /etc/os-release):  Ubuntu 16.04 LTS (Xenial Xerus)
- **Kernel** (e.g. `uname -a`): Linux k8s-master-18117938-0 4.4.0-28-generic #47-Ubuntu SMP Fri Jun 24 10:09:13 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
- **Install tools**: Azure ACS-Engine
- **Others**: 

**What happened**:

I have a service of `type: LoadBalancer`. I specify `loadBalancerSourceRanges` in the spec. On initial create of this service (via `kubectl apply`, `loadBalancerSourceRanges` is correctly used to define the Azure NSG. However, on update of this service (again via `kubectl apply`), `loadBalancerSourceRanges` in the service definition is updated (as can be seen on the dashboard), but the associated Azure NSG inbound rules do not update.

If the service is deleted and recreated, the NSG is deleted and created correctly.

**What you expected to happen**:

The changes in `loadBalancerSourceRanges` should be applied to the Azure NSG.

**How to reproduce it** (as minimally and precisely as possible):

* Create a service of type `LoadBalancer` via `kubectl apply -f <file>`, similar to:
```
apiVersion: v1
kind: Service
metadata:
  name: ingress-controller-external
  namespace: kube-system
  labels:
    k8s-app: ingress-controller-external
  annotations:
    # For Source IP support (see https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-typeloadbalancer)
    service.beta.kubernetes.io/external-traffic: OnlyLocal
spec:
  type: LoadBalancer
  loadBalancerSourceRanges:
  - 24.150.0.0/16
  ports:
  - name: http
    port: 80
    targetPort: 80
  - name: https
    port: 443
    targetPort: 443
  - name: health
    port: 10254
    targetPort: 10254
  selector:
    k8s-app: ingress-controller-external
```

* Wait for it to deploy
* Edit the yaml and modify the `loadBalancerSourceRanges`.
* `kubectl apply -f <file>`
* Note the change is not applied on the Azure NSG

**Anything else we need to know**:

This is similar to issue #34218 but the difference here is that changing an annotation in order to apply the update *does not* fix/work around the problem.

Originally reported @ https://github.com/Azure/acs-engine/issues/423 and tagging @brendandburns  as requested by @colemickens .

**Comment by karataliu** [comment]
Same cause: https://github.com/kubernetes/kubernetes/issues/43691

**Comment by karataliu** [comment]
Related PR #45523 was merged to v1.7, and has been cherry-picked to v1.6

Verified with v1.6.6, this issue is fixed.

**Comment by itowlson** [comment]
Should this now be closed?

**Comment by feiskyer** [comment]
Have been fixed in #45523. Closing this issue.
